ignite/distributed/utils.py

```diff
@@ -0,0 +1,220 @@
+        assert dist.is_available() and dist.is_initialized()
+        return self._local_rank
+
+    @staticmethod
+    def get_rank() -> int:
+        assert dist.is_available() and dist.is_initialized()
+        return dist.get_rank()
+
+    @staticmethod
+    def get_world_size() -> int:
+        assert dist.is_available() and dist.is_initialized()
+        return dist.get_world_size()
+
+    @staticmethod
+    def is_distributed() -> bool:
+        return dist.is_available() and dist.is_initialized() and dist.get_world_size() > 0
+
+    @staticmethod
+    def name() -> str:
+        return "dist"
+
+
+# default is sequential model
+_model = SequentialModel()
+# default backend is None
+_backend = None
+# default device is None
+_device = None
+# the model, device and backend could change at initialization
+
+
+def is_dist_distributed() -> bool:
+    return _model.is_distributed() and _model.name() != "xla"
+
+
+def is_tpu_distributed() -> bool:
+    return _model.is_distributed() and _model.name() == "xla"
+
+
+def get_world_size() -> int:
+    return _model.get_world_size()
+
+
+def get_rank() -> int:
+    return _model.get_rank()
+
+
+def get_local_rank() -> int:
+    return _model.get_local_rank()
+
+
+def get_device():
+    return _device
+
+
+def get_backend():
+    return _backend
+
+
+def get_hostname() -> str:
+    return socket.gethostname()
+
+
+def _find_best_dist_backend():
+    assert dist.is_available()
+    # select best backend nccl first, gloo second
+    if dist.is_nccl_available():
+        return "nccl"
+    elif dist.is_gloo_available():
+        return "gloo"
+    else:
+        raise Exception("No backend found (missing gloo and nccl")
+
+
+def _find_best_dist_device():
+    # select best device cuda is possible otherwise cpu
+    if torch.cuda.is_available() and torch.cuda.device_count() > 0:
+        return "cuda"
+    else:
+        return "cpu"
+
+
+def initialize(backend=None, device=None):
+    global _model, _backend, _device
+    if has_xla_support:
+        # no backend for tpu
+        assert backend is None
+        # if device imposed, it should be tpu
+        if device is not None:
+            assert device == "tpu"
+        _device = "tpu"
+        _model = XlaModel()
+        # how to check if already initialized ?
+        xm.rendezvous("init")
+    elif dist.is_available():
+        if backend is None:
+            _backend = _find_best_dist_backend()
+        else:
+            # imposed by user
+            _backend = backend
+        assert not dist.is_initialized()
+        if backend == "nccl":
+            torch.backends.cudnn.benchmark = True
+        if device is None:
+            _device = _find_best_dist_device()
+        else:
+            _device = device
+        _model = DistModel()
+        dist.init_process_group(backend, init_method="env://")
+        if _device == "cuda":
+            torch.cuda.set_device(get_local_rank())
+        # TODO: check if backend is available
```

Comment:
If it is init then it is available ?

Reply:
I agree, it's buggy if dist not available. I must use a docker with a torch without dist... (dont know what it means, maybe old pytorch ?)